backend/controller/cronjobs/cronjobs.go

```diff
@@ -0,0 +1,483 @@
+
+	for {
+		sl.SortFunc(state.jobs, func(i, j dal.CronJob) int {
+			return s.compareJobs(state, i, j)
+		})
+
+		now := s.clock.Now()
+		next := now.Add(time.Hour) // should never be reached, expect a different signal long beforehand
+		for _, j := range state.jobs {
+			if possibleNext, err := s.nextAttemptForJob(j, state, false); err == nil {
+				next = possibleNext
+				break
+			}
+		}
+
+		if next.Before(state.blockedUntil) {
+			next = state.blockedUntil
+			logger.Tracef("loop blocked for %v", next.Sub(now))
+		} else if next.Sub(now) < time.Second {
+			next = now.Add(time.Second)
+			logger.Tracef("loop while gated for 1s")
+		} else if next.Sub(now) > time.Minute*59 {
+			logger.Tracef("loop while idling")
+		} else {
+			logger.Tracef("loop with next %v, %d jobs", next.Sub(now), len(state.jobs))
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-s.clock.After(next.Sub(now)):
+			// Try starting jobs in db
+			jobsToAttempt := slices.Filter(state.jobs, func(j dal.CronJob) bool {
+				if n, err := s.nextAttemptForJob(j, state, true); err == nil {
+					return !n.After(s.clock.Now().UTC())
+				}
+				return false
+			})
+			jobResults, err := s.dal.StartCronJobs(ctx, jobsToAttempt)
+			if err != nil {
+				logger.Errorf(err, "failed to start cron jobs in db")
+				state.blockedUntil = s.clock.Now().Add(time.Second * 5)
+				continue
+			}
+
+			// Start jobs that were successfully updated
+			updatedJobs := []dal.CronJob{}
+			removedDeploymentKeys := map[string]model.DeploymentKey{}
+
+			for _, job := range jobResults {
+				updatedJobs = append(updatedJobs, job.CronJob)
+				if !job.DidStartExecution {
+					continue
+				}
+				if !job.HasMinReplicas {
+					// We successfully updated the db to start this job but the deployment has min replicas set to 0
+					// We need to update the db to end this job
+					removedDeploymentKeys[job.DeploymentKey.String()] = job.DeploymentKey
+					_, err := s.dal.EndCronJob(ctx, job.CronJob, next)
+					if err != nil {
+						logger.Errorf(err, "failed to end cron job %s", job.Key.String())
+					}
+					continue
+				}
+				logger.Infof("executing job %v", job.Key)
+				state.startedExecutingJob(job.CronJob)
+				go s.executeJob(ctx, job.CronJob)
+			}
+
+			// Update job list
+			state.updateJobs(updatedJobs)
+			for _, key := range removedDeploymentKeys {
+				state.removeDeploymentKey(key)
+			}
+		case event := <-jobChanges:
+			switch event.changeType {
+			case resetJobs:
+				logger.Tracef("resetting job list: %d jobs", len(event.jobs))
+				state.reset(event.jobs, event.addedDeploymentKey)
+			case finishedJobs:
+				logger.Tracef("updating %d jobs", len(event.jobs))
+				state.updateJobs(event.jobs)
+			case updatedHashring:
+				// do another cycle through the loop to see if new jobs need to be scheduled
+			}
+		}
+	}
+}
+
+func (s *Service) compareJobs(state *State, i, j dal.CronJob) int {
+	iNext, err := s.nextAttemptForJob(i, state, false)
+	if err != nil {
+		return 1
```

Comment:
This is a bit of a code smell...

Reply:
Removed, sorting wasn't necessary in the end